src/RelayProvider.ts

```diff
@@ -0,0 +1,278 @@
+            return true;
+        }
+
+        return false;
+    }
+
+    private _getRelayStatus(respResult: TransactionReceipt): {
+        relayRevertedOnRecipient: boolean;
+        transactionRelayed: boolean;
+        reason?: string;
+    } {
+        if (!respResult.logs || respResult.logs.length === 0) {
+            return {
+                relayRevertedOnRecipient: false,
+                transactionRelayed: false,
+                reason: 'Tx logs not found'
+            };
+        }
+
+        const relayHubInterface = RelayHub__factory.createInterface();
+        const parsedLogs = respResult.logs.map(log => relayHubInterface.parseLog(log));
+
+        const recipientRejectedEvents = parsedLogs.find(log => log.name == 'TransactionRelayedButRevertedByRecipient');
+
+        if (recipientRejectedEvents) {
+            const { reason } = recipientRejectedEvents.args;
+
+            const revertReason = (reason as string) ?? 'Unknown';
+            log.info(
+                `Recipient rejected on-chain: ${revertReason}`
+            );
+
+            return {
+                relayRevertedOnRecipient: true,
+                transactionRelayed: false,
+                reason: revertReason
+            };
+        }
+
+        const transactionRelayed = parsedLogs.find(log => log.name == 'TransactionRelayed');
+
+        if (transactionRelayed) {
+            return {
+                relayRevertedOnRecipient: false,
+                transactionRelayed: true
+            };
+        }
+
+        // Check if it wasn't a deploy call which does not emit TransactionRelayed events but Deployed events
+        const deployedEvent = parsedLogs.find(log => log.name == 'Deployed');
+
+        if (deployedEvent) {
+            return {
+                relayRevertedOnRecipient: false,
+                transactionRelayed: true
+            };
+        }
+
+        log.info(
+            'Neither TransactionRelayed, Deployed, nor TransactionRelayedButRevertedByRecipient events found. This might be a non-enveloping transaction '
+        );
+
+        return {
+            relayRevertedOnRecipient: false,
+            transactionRelayed: false,
+            reason: 'Neither TransactionRelayed, Deployed, nor TransactionRelayedButRevertedByRecipient events found. This might be a non-enveloping transaction'
+        };
+    }
+
+    private async executeRelayTransaction(
+        envelopingRequest: UserDefinedEnvelopingRequest,
+        requestConfig: RequestConfig
+    ): Promise<RelayingResult> {
+        try {
+
+            const transaction = await this.relayClient.relayTransaction(
+                envelopingRequest, requestConfig
+            );
+
+            const { hash } = transaction;
+
+            if(!hash){
+                throw new Error('Transaction has no hash!');
+            }
+
+            let receipt;
+            if(!requestConfig.ignoreTransactionReceipt){
+                receipt = await this.getTransactionReceipt(hash);
```

Comment:
I think that ethers has a special method, to wait for the transaction. 
https://docs.ethers.org/v5/api/providers/provider/#Provider-waitForTransaction

Reply:
very nice find.